kore/test/Test/Kore/Step/Simplification/SubstitutionSimplifier.hs

```diff
@@ -0,0 +1,231 @@
+{-# LANGUAGE OverloadedLists #-}
+
+module Test.Kore.Step.Simplification.SubstitutionSimplifier
+    ( test_SubstitutionSimplifier
+    ) where
+
+import Test.Tasty
+
+import qualified GHC.Stack as GHC
+
+import qualified Kore.Internal.OrPredicate as OrPredicate
+import qualified Kore.Internal.Predicate as Predicate
+import Kore.Internal.TermLike
+import Kore.Step.Simplification.SubstitutionSimplifier
+    ( SubstitutionSimplifier (..)
+    )
+import qualified Kore.Step.Simplification.SubstitutionSimplifier as SubstitutionSimplifier
+import Kore.Unification.Error
+    ( SubstitutionError (..)
+    , UnificationOrSubstitutionError (..)
+    )
+import qualified Kore.Unification.Substitution as Substitution
+import Kore.Unification.SubstitutionNormalization
+import Kore.Unification.Unify
+    ( runUnifierT
+    )
+import Kore.Variables.UnifiedVariable
+    ( UnifiedVariable (..)
+    )
+
+import qualified Test.Kore.Step.MockSymbols as Mock
+import Test.Kore.Step.Simplification
+    ( runSimplifier
+    )
+import Test.Tasty.HUnit.Ext
+
+test_SubstitutionSimplifier :: [TestTree]
+test_SubstitutionSimplifier =
+    [ test "empty substitution"
+        []
+        [ mempty ]
+    , test "normalized substitution"
+        [(y, a)]
+        [ Normalization [(y, a)] [] ]
+    , test "unnormalized substitution, variable-only"
+        [(y, mkVar x), (x, a)]
+        [ Normalization [(x, a), (y, a)] [] ]
+    , test "unnormalized substitution, variable under symbol"
+        [(y, sigma (mkVar x) b), (x, a)]
+        [ Normalization [(x, a), (y, sigma a b)] [] ]
+    , testGroup "element-variable-only cycle"
+        [ test "length 1, alone"
+            [(x, mkVar x)]
+            [ mempty ]
+        , test "length 1, beside related substitution"
+            [(x, mkVar x), (z, f (mkVar x))]
+            [ Normalization [(z, f (mkVar x))] [] ]
+        , test "length 1, beside unrelated substitution"
+            [(x, mkVar x), (z, a)]
+            [ Normalization [(z, a)] [] ]
+        ]
+    , testGroup "set-variable-only cycle"
+        [ test "length 1, alone"
+            [(xs, mkVar xs)]
+            [ mempty ]
+        -- UnificationError: UnsupportedPatterns
+        -- , test "length 1, beside related substitution"
+        --     [(xs, mkVar xs), (ys, mkVar xs)]
+        --     [([(ys, mkVar xs)], [])]
```

Comment:
Should this remain commented?

Reply:
Yes; this test has an error due to a missing unification case. I do not know if we will ever add that unification case, but if we do then the test should be uncommented. I will add a TODO item for that.